pkg/publicapi/apimodels/error.go

```diff
@@ -0,0 +1,131 @@
+package apimodels
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+// apierror represents a standardized error response for the api.
+//
+// it encapsulates:
+//   - an http status code
+//   - a human-readable error message
+//
+// purpose:
+//   - primarily used to send synchronous errors from the orchestrator endpoint
+//   - provides a structured json error response to http clients
+//
+// usage:
+//   - when the cli interacts with an orchestrator node via an http client:
+//     1. the http client receives this structured json error
+//     2. the human-readable message is output to stdout
+//     3. the http status code allows for programmatic handling of different error types
+//
+// benefits:
+//   - ensures consistent error formatting across the api
+//   - facilitates both user-friendly error messages and machine-readable error codes
+type APIError struct {
+	// httpstatuscode is the http status code associated with this error.
+	// it should correspond to standard http status codes (e.g., 400, 404, 500).
+	HTTPStatusCode int `json:"code"`
+
+	// message is a short, human-readable description of the error.
+	// it should be concise and provide a clear indication of what went wrong.
+	Message string `json:"message"`
+}
```

Comment:
I have added the RequestID. However, adding a bacalhau specific error code is tricky at this point. Mainly because we are need to make sure that we properly propagate error codes correctly every where. I think we should pick that up in follow ups. Let me know what you think

Reply:
Yeah please feel free to do this in a follow up PR. What I had in mind is something similar to the code below. The idea is to have standard error codes that can be used by our core components and even the modules if the errors apply, while still enabling the modules to define their own specific error code, such as `BucketNotFound` and `ContainerStartFailed`

The `HTTPCode` can be explicitly provided, inferred from the error code if it is a standard one, or fallback to 500.

What I really liked in your PR is the introduction of `Component` field which helps identify the source of the error and also removes the need for these error codes to be globally unique. I would just make the components more readable and descriptive, such as `S3Publisher`, `Docker` instead of abbreviations. To simplify things, I am proposing the default component to be `Bacalhau`, which doesn't mean much, but will allow us to incrementally improve the coverage of better errors.

Also as we progress, we should allow components to override the `Component` piece of the errors. Meaning a job not found error with `JobStore` component doesn't help much compared to knowing the component that actually called the job store. We don't need to do these changes today, but just thinking ahead.

What do you think?

```go
const (
    BadRequest ErrorCode = "BadRequest"
    Unauthorized ErrorCode = "Unauthorized"
    ResourceNotFound ErrorCode = "ResourceNotFound"
    ResourceInUse ErrorCode = "ResourceInUse"
    ValidationFailed ErrorCode = "ValidationFailed"
    VersionMismatch ErrorCode = "VersionMismatch"
    TooManyRequests ErrorCode = "TooManyRequests"
    
    InternalError ErrorCode = "InternalError"
    NotImplemented ErrorCode = "NotImplemented"
    ServiceUnavailable ErrorCode = "ServiceUnavailable"
    DatastoreFailure ErrorCode = "DatastoreFailure"
    NetworkFailure ErrorCode = "NetworkFailure"
    ConfigurationError ErrorCode = "ConfigurationError"
    ResourceExhausted ErrorCode = "ResourceExhausted"
)

// Note: Component-specific error codes should be defined in their respective packages.
// Examples might include:
// - BucketNotFound
// - ObjectNotFound
// - InvalidCredentials
// - ImageNotFound
// - ContainerStartFailed
// - PeerConnectionFailed

type BaseError struct {
    Code           ErrorCode         `json:"code"`
    Message        string            `json:"message"`
    Component      string            `json:"component"`
    HTTPCode       int               `json:"http_code"`
    Hint           string            `json:"hint,omitempty"`
    Retryable      bool              `json:"retryable"`
    FailsExecution bool              `json:"fails_execution"`
    Details        map[string]string `json:"details,omitempty"`
}

func NewBaseError(format string, a ...any) *BaseError {
    return &BaseError{
        Code:      InternalError,
        Message:   fmt.Sprintf(format, a...),
        Component: "Bacalhau",
        HTTPCode:  0, // Will be inferred later if not set explicitly
    }
}

func (e *BaseError) Error() string {
    return fmt.Sprintf("[%s] %s: %s", e.Code, e.Component, e.Message)
}

func (e *BaseError) WithErrorCode(code ErrorCode) *BaseError {
    e.Code = code
    return e
}

func (e *BaseError) WithHTTPCode(httpCode int) *BaseError {
    e.HTTPCode = httpCode
    return e
}

func (e *BaseError) HTTPStatusCode() int {
    if e.HTTPCode != 0 {
        return e.HTTPCode
    }
    return inferHTTPStatusCode(e.Code)
}

func inferHTTPStatusCode(code ErrorCode) int {
    switch code {
    case BadRequest, ValidationFailed:
        return http.StatusBadRequest
    case Unauthorized:
        return http.StatusUnauthorized
    case ResourceNotFound:
        return http.StatusNotFound
    case ResourceInUse, VersionMismatch:
        return http.StatusConflict
    case TooManyRequests:
        return http.StatusTooManyRequests
    case NotImplemented:
        return http.StatusNotImplemented
    case ServiceUnavailable:
        return http.StatusServiceUnavailable
    default:
        return http.StatusInternalServerError
    }
}

// rest of the methods
```